drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/widget/FactPatternWidgetTest.java

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.drools.workbench.screens.guided.rule.client.widget;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import com.google.gwt.event.shared.EventBus;
+import com.google.gwt.user.client.ui.FlexTable;
+import com.google.gwtmockito.GwtMockitoTestRunner;
+import com.google.gwtmockito.WithClassesToStub;
+import org.appformer.project.datamodel.oracle.DataType;
+import org.appformer.project.datamodel.oracle.FieldAccessorsAndMutators;
+import org.appformer.project.datamodel.oracle.ModelField;
+import org.appformer.project.datamodel.oracle.OperatorsOracle;
+import org.drools.workbench.models.datamodel.rule.FactPattern;
+import org.drools.workbench.models.datamodel.rule.FieldConstraint;
+import org.drools.workbench.models.datamodel.rule.SingleFieldConstraint;
+import org.drools.workbench.screens.guided.rule.client.editor.CEPOperatorsDropdown;
+import org.drools.workbench.screens.guided.rule.client.editor.RuleModeller;
+import org.drools.workbench.screens.guided.rule.client.editor.factPattern.Connectives;
+import org.drools.workbench.screens.guided.rule.client.resources.images.GuidedRuleEditorImages508;
+import org.jboss.errai.common.client.api.Caller;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.kie.workbench.common.services.datamodel.service.IncrementalDataModelService;
+import org.kie.workbench.common.services.shared.preferences.ApplicationPreferences;
+import org.kie.workbench.common.widgets.client.datamodel.AsyncPackageDataModelOracleImpl;
+import org.kie.workbench.common.widgets.client.datamodel.OracleUtils;
+import org.mockito.Mock;
+import org.uberfire.mocks.CallerMock;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+
+@WithClassesToStub({FlexTable.class, GuidedRuleEditorImages508.class, CEPOperatorsDropdown.class})
+@RunWith(GwtMockitoTestRunner.class)
+public class FactPatternWidgetTest {
+
+    private Caller<IncrementalDataModelService> serviceCaller;
+
+    @Mock
+    private CEPOperatorsDropdown operatorsDropdown;
+
+    @Mock
+    private IncrementalDataModelService service;
+
+    @Mock
+    private AsyncPackageDataModelOracleImpl oracle;
+
+    @Mock
+    private RuleModeller ruleModeller;
+
+    @Mock
+    private EventBus eventBus;
+
+    @Mock
+    private FactPattern pattern;
+
+    @Mock
+    private SingleFieldConstraint fieldConstraint;
+
+    private boolean canBind = true;
+
+    private boolean readOnly = false;
+
+    private FactPatternWidget factPatternWidget;
+
+    private Connectives connectives;
+
+    @Before
+    public void setUp() throws Exception {
```

Comment:
There's a lot of duplication in this test's setup compared to ```ConnectivesTest```.. any chance of refactoring shared aspects to a common test class?

Reply:
Challenge accepted